### DISCUSSION TITLE
VS Code Notebooks

### DISCUSSION THREAD

**Discussion by JanikNex** [eclipse-langium/langium#1403] [Q&A] [answered]
Hey everyone,

I am currently trying to implement notebook support for a Langium language in VSCode.
I have followed https://www.typefox.io/blog/langium-notebooks/ and https://github.com/TypeFox/langium-lox, but I am currently facing problems and don't really know how to fix them. I am pretty sure that I have followed all the steps according to the blog entry.

**General background:**
Two Langium languages already exist in my project (Langium v2.1.3), now I want to introduce a third language which can be used within a notebook.
The new language has the file extension `.gm`, the notebook the file extension `.gmnb`.

**The problem:**
The language itself works perfectly, but when I load the extension and create a code cell in a new notebook, I get the following error messages:

```
Error: Error: The service registry contains no services for the extension '.gmnb'.
	at DefaultServiceRegistry.getServices (c:\Users\USER\.vscode\extensions\mml-0.0.2\out\language\main.cjs:32444:13)
	at DefaultLangiumDocumentFactory.parse (c:\Users\USER\.vscode\extensions\mml-0.0.2\out\language\main.cjs:20326:43)
	at DefaultLangiumDocumentFactory.create (c:\Users\USER\.vscode\extensions\mml-0.0.2\out\language\main.cjs:20265:32)
	at DefaultLangiumDocuments.getOrCreateDocument (c:\Users\USER\.vscode\extensions\mml-0.0.2\out\language\main.cjs:20358:46)
	at DefaultDocumentBuilder.update (c:\Users\USER\.vscode\extensions\mml-0.0.2\out\language\main.cjs:32689:31)
	at c:\Users\USER\.vscode\extensions\mml-0.0.2\out\language\main.cjs:20559:43
	at c:\Users\USER\.vscode\extensions\mml-0.0.2\out\language\main.cjs:14637:65

[Error - 12:57:30 AM] Request textDocument/semanticTokens/full failed.
	Message: Request textDocument/semanticTokens/full failed with message: The service registry contains no services for the extension '.gmnb'.
	Code: -32603
```

As a possible solution, I then added the notebook's file extension (`.gmnb`) to the GM language in `langium-config.json` as an additional file extension.

In VSCode, the code cells now load correctly and offer the expected functions of the language server.

However, after the notebook serializer (simple JSON serializer) has saved the whole notebook to the file - VSCode now displays error messages, from which I can conclude that the parser now also considers the serialized notebook as JSON (which of course it cannot parse as a GM language document).

<img width="529" alt="gmnb_errors" src="https://github.com/eclipse-langium/langium/assets/18660622/4d24d743-3bcd-41e7-a0ee-1b03f8962542">

The notebook still works as expected and without problems, but VSCode displays those parser errors.

I am therefore not sure where the problem lies. Have I overlooked something, or has something fundamental changed in Langium or VSCode since the blog post, so that the steps are no longer sufficient?

Has anyone had any experience with Langium in notebooks in more recent versions?

**Comment by msujew** [comment, marked as answer]
Hey @JanikNex, sorry for the late response.

the main issue comes from the fact that you have two languages already and the service registry isn't capable of selecting the correct one for the notebook cell documents. This issue isn't exhibited by language servers that only support a single language, as the service registry doesn't need to pick - it will always return the singleton service instance for any document. The example I'm using in the blog post is based on lox, which is just a single language.

The workaround you proposed has the issue that the language server will attempt to load the full `.gmnb` file as a document of your language. Since this isn't the case (it's just a json file), you just get a bunch of parser errors. So this isn't really the way to go.

The correct way to solve this is to create a specialized [`ServiceRegistry`](https://github.com/eclipse-langium/langium/blob/main/packages/langium/src/service-registry.ts) implementation. This new `ServiceRegistry` should return the same `LangiumServices` instance for any `.gmnb` file as for the normal `.gm` files.

**Reply by JanikNex** [reply]
Hey, thanks for the answer!

> The workaround you proposed has the issue that the language server will attempt to load the full .gmnb file as a document of your language

Yeah, I noticed that then, but at first it was the most obvious way to add another file type.

> The correct way to solve this is to create a specialized [ServiceRegistry](https://github.com/eclipse-langium/langium/blob/main/packages/langium/src/service-registry.ts) implementation.

Ah! I wasn't aware that the `DefaultServiceRegistry` distinguishes between single-language and multi-language projects. I looked for the cause of the error in various places, but not in the `ServiceRegistry` itself (my bad).

I have now overridden `getServices()` in a specialized `ServiceRegistry` as suggested and now it works. 

Thank you!
